### DISCUSSION TITLE
Post 3.0 Roadmap

### DISCUSSION THREAD

**Discussion by starseeker** [elalish/manifold#1064] [General]
@elalish now that 3.0 has been released, what are the next priorities?

**Comment by elalish** [comment]
Well, I suppose I should turn that question around - what would you like to contribute? I'm certainly hoping we can stay away from breaking changes for awhile now. There are lots of things I'd like to do, but my time is pretty constrained. As usual, bugs get top priority. This is my personal list:

- #1017/#1040 are my current focus. I'm hoping that turns into a solid generic mesh decimator.
- #970/#1006/#666; it seems like offsetting is a good edge case test that's catching some Boolean issue. Need to investigate.
- #972 I think could bring a significant perf improvement and reduce lines of code. I'm probably the only person with the background for this particular refactor, as it will be large.
- #289 is the big new algorithm I want to write that would open up a lot of use cases. I'll need to find a serious chunk of time to tackle that.
- I wrote a novel straight skeleton algorithm years ago I've been meaning to turn into C++: it turns a polygon into a manifold 3D "roof". If anyone thinks that's useful, I might prioritize it higher. 

I'm curious what @pca006132 wants to tackle - I bet it involves perf improvements. 

What does everyone else want? What might you like to contribute? 
- There's much talk of Rust bindings, e.g. #1046. 
- It'd be nice to release the worker of ManifoldCAD.org as its own npm package. 
- It would be especially cool to then use that package to make a VSCode extension that's effectively ManifoldCAD.org, but with file system support. Especially if that made it easy to import and export design libraries like OpenSCAD does. That would be an awesome project for someone who wants to help, but not get into the nitty-gritty of geometric algorithms. 
- I wrote another glTF [extension](https://github.com/KhronosGroup/glTF/pull/2458) (just a draft) to encode our smoothing tangents. It would be cool to make a lighter-weight JS decoder for that to demonstrate the benefits of local tessellation. Anyone interested?

**Comment by pca006132** [comment]
My plan:

1. ~~Add debug code to dump the boolean operands that produce an error in `csg_tree`, to help debugging things like #970 and #1006.~~
2. ~~Fix #989, I have some idea about how to fix it, but it will be a big refactor for `csg_tree`, due to c++ having no support for tail recursion...~~
3. Look into #982, I think we have some low-hanging fruits there, but that may require some refactoring. And there are potential algorithmic improvements as well.
4. Compile manifold to wasm without emscripten. I think it will take a long time until the rust folks add wasm-unknown-emscripten to wasm-bindgen, so it will probably be easier to support compiling manifold to wasm-unknown-unknown directly. Requires some hacking, but I think this will be a fun one.
5. Better SDF: Look into providing a custom IR for user to build, similar to libfive.

**Comment by starseeker** [comment]
@elalish Saw you've switched jobs to working at Wētā - just for planning purposes, is that likely to push out any investigation into the potential boolean bug?

**Reply by elalish** [reply]
Yeah, I just moved my family of 5 to a different hemisphere - I'm pretty slammed with logistics at the moment. I'd love to carve out some time for Manifold again, but realistically it'll probably be at least another month before that happens. I can manage issue responses and code reviews, so it would be great to see some other contributors step up to write some code, but we'll see how it goes.

**Reply by starseeker** [reply]
Oh, wow!  That's a huge change - so it's not remote work but an actual physical relocation.  Wow!  I've moved before, but never on that scale - best of luck!

**Comment by napter** [comment]
+1 vote for overlap removal! I believe that should allow for sweeps and other self-intersecting operations.

**Comment by ChillstepCoder** [comment]
+1 for straight skeleton algorithm! Would love to dump CGal for our procedural roof generation :)

**Comment by mmiscool** [comment]
I would be very interested in a sweep operation.

**Comment by briansturgill** [comment]
I'm working on fillet/chamfer in my [Piecad](https://www.github.com/briansturgil/Piecad) project and ran into an interesting
problem with rotate.   To make a long story short, I had to use quaternion rotation to make things place properly.
I know very little about quaternions, but in reading up on them, it seems that many game people and CAD people are
not using Transformation matrices (more memory, more ops) but rather a rotation quaternion, and two vectors: scale and translation.
There is much information about this, for example [on Wikipedia](https://en.wikipedia.org/wiki/Quaternions_and_spatial_rotation).

**Reply by briansturgill** [reply]
Here's something from Grok: 
[grok_report.pdf](https://github.com/user-attachments/files/32547369/grok_report.pdf)

**Reply by elalish** [reply]
Yes, I love quaternions, definitely use them! We use them internally a bit already. If you're referring to our API, our 3x4 transform is a superset of quaternions, scale, and translation, so it works fine - you just need a conversion utility (I'd recommend linalg.h like we use). Also, beware of vector scale - I'd recommend keeping scale as a scalar if you use that approach, otherwise a series of operations is not representable in that same compact form, though it will be with a full 3x4 matrix.
